src/librustc/ty/print/pretty.rs

```diff
@@ -1532,12 +1536,95 @@ define_print_and_forward_display! {
     }
 
     &'tcx ty::Const<'tcx> {
-        match self.val {
-            ConstValue::Unevaluated(..) |
-            ConstValue::Infer(..) => p!(write("_")),
-            ConstValue::Param(ParamConst { name, .. }) => p!(write("{}", name)),
-            _ => p!(write("{:?}", self)),
+        let u8 = cx.tcx().types.u8;
+        if let ty::FnDef(did, substs) = self.ty.sty {
+            p!(print_value_path(did, substs));
+            return Ok(cx);
+        }
+        if let ConstValue::Unevaluated(did, substs) = self.val {
+            match cx.tcx().describe_def(did) {
+                | Some(Def::Static(_, _))
+                | Some(Def::Const(_, false))
+                | Some(Def::AssociatedConst(_)) => p!(write("{}", cx.tcx().def_path_str(did))),
```

Comment:
This does the same thing as `print_value_path`, except less ideal (printing shouldn't call `def_path_str`, ever, since `def_path_str` is a wrapper around printing without `substs`).